primary_beam_correction.py

```diff
@@ -0,0 +1,296 @@
+    raw_img = []
+    try:
+        nterm = raw_image.header['NTERM']
+    except KeyError:
+        nterm = _get_value_from_history('NTERM', raw_image.header)
+    nterm = int(nterm)
+    for j in range(nterm, raw_image.header['NSPEC']+nterm):
+        raw_img.append(np.ravel(raw_image.data[0, j, :, :]))
+    snr = []
+    pbc_image = []
+    for i in range(len((beam_pattern))):
+        # Getting all the values with attenuated flux of less than 10% of the peak
+        ind = np.argwhere(beam_pattern[i] <= 0.10)
+        if len(ind) > 0:
+            beam_pattern[i][ind] = np.nan
+        # Getting the rms noise in each image (before primary beam correction)
+        snr.append(standard_deviation(raw_img[i]))
+        # To correct the effect of the beam we divide by the beam.
+        ratio = raw_img[i] / beam_pattern[i]
+        pbc_image.append(ratio)
+    # Primary beam corrected (pbc) image
+    pbc_image = np.array(pbc_image)
+    snr = np.array(snr)
+    # Calculating a weighted average of the indidvidual frequency lane images
+    corr_image = weigted_average(pbc_image, snr)
+    corr_image = corr_image.reshape(raw_image.data.shape[2], raw_image.data.shape[3])
+    # Add new axis
+    corr_image = corr_image[np.newaxis, np.newaxis, :, :]
+    return corr_image
+
+
+def _get_value_from_history(keyword, header):
+    ''''
+    Return the value of a keyword from the FITS HISTORY in header
+    Assumes keyword is found in a line of the HISTORY with format: 'keyword = value'
+    Input: Keyword [e.g BMAJ, ClEANBMJ, BMIN] and Image header
+    '''
+    for history in header['HISTORY']:
+        line = history.replace('=', ' ').split()
+        try:
+            ind = line.index(keyword)
+        except ValueError:
+            continue
+        return line[ind + 1]
+    raise KeyError(f'{keyword} not found in HISTORY') 
+    
+
+def write_new_fits(pbc_image, path, outputFilename):
+    '''
+    Write out a new fits image with primary beam corrected continuum in its first  plane
+    '''
+    images = read_fits(path)
+    hdr = images.header
+    newhdr = hdr.copy()
+    # change the frequency plane keywords, we don't want multiple frequency axes
+    newhdr['CTYPE3'] = 'FREQ'
+    newhdr['NAXIS3'] = 1
+    newhdr['CDELT3'] = 1.0
+    try:
+        if newhdr['CLEANBMJ'] > 0:
+            # add in required beam keywords
+            newhdr['BMAJ'] = newhdr['CLEANBMJ']
+            newhdr['BMIN'] = newhdr['CLEANBMN']
+            newhdr['BPA'] = newhdr['CLEANBPA']
+        else:
+            # Checking CLEANBMAJ in the history 
+            newhdr['BMAJ'] = float(_get_value_from_history('BMAJ', newhdr))
+            newhdr['BMIN'] = float(_get_value_from_history('BMIN', newhdr))
+            newhdr['BPA'] = float(_get_value_from_history('BPA', newhdr))
+    except KeyError:
+        logging.error('Exception occurred, keywords not found', exc_info=True)
+    data_new = pbc_image[:, 0:1, :, :]  # the shape of data_new = (1,1, npix, npix)
```

Comment:
Is the slice necessary?

Reply:
Not at all, even that line itself is not necessary, I fixed it.